### ISSUE TITLE
no pcmax implementation in bd2kccd/r-causal

### ISSUE THREAD

**Issue by kingfish777** [bd2kccd/r-causal#28]
Greetings,
Please put a means to implement pcmax in r-causal. I am performing randomized sub-sampling of millions of rows/observations (and between 20 to 30K+ discrete binary variables)  and I am going to do an alpha = .000005. I will be looking for stable edges across samples. 
Thanks!
Scott

**Comment by kingfish777** [comment]
... which is strange because I see it here: https://github.com/bd2kccd/r-causal/blob/master/java/edu/cmu/tetrad/search/PcMax.java

**Comment by espinoj** [comment]
We've seen instances where the a recent R library is not loaded unless the old one is explicitly unloaded.  You may want to try that.

See this thread
https://github.com/bd2kccd/r-causal/issues/23

**Comment by kingfish777** [comment]
Nope. Still not working. Detached, removed rcausal installations, library(devtools)/install_github'd, still no success (in seeing pcmax).
 -S

**Comment by chirayukong** [comment]
Scott,

We don't currently support pcmax in the r-causal library yet. I'll put it in the next release. I'll notify you when it's done. In the mean time, you can use pcmax in the Tetrad application http://www.phil.cmu.edu/tetrad/.

Regards,

Kong

**Comment by kingfish777** [comment]
Hi Chiyaru,

That's alright. I went ahead and wrote a wrapper for it and am debugging it (of course, in order to test it, NAMESPACE and INDEX in main folder have to be updated). I am testing it now.

 --S


############




pcmax <- function(df, continuous = TRUE, depth = 3,
        #aggressivelyPreventCycles = FALSE,
    falseDiscoveryRate = FALSE, significance = 0.05, #maxPathLength = NULL,
    #useHeuristic = FALSE,
    verbose = FALSE, java.parameters = NULL, priorKnowledge = NULL) {

    params <- list(NULL)

    if(!is.null(java.parameters)){
        options(java.parameters = java.parameters)
        params <- c(java.parameters = java.parameters)
    }

    # Data Frame to Independence Test
    indTest = NULL
    if(continuous){
        tetradData <- loadContinuousData(df)
        indTest <- .jnew("edu/cmu/tetrad/search/IndTestFisherZ", tetradData, significance)
    }else{
        tetradData <- loadDiscreteData(df)
        indTest <- .jnew("edu/cmu/tetrad/search/IndTestChiSquare", tetradData,
            significance)
    }

    indTest <- .jcast(indTest, "edu/cmu/tetrad/search/IndependenceTest")

    pcmax <- list()
    class(pcmax) <- "pcmax"

    pcmax$datasets <- deparse(substitute(df))

    cat("Datasets:\n")
    cat(deparse(substitute(df)),"\n\n")


    # Initiate PC
    pcmax_instance <- .jnew("edu/cmu/tetrad/search/PcMax", indTest)
    .jcall(pcmax_instance, "V", "setDepth", as.integer(depth))
    #.jcall(pcmax_instance, "V", "setMaxPathLength", as.integer(maxPathLength))
   # .jcall(pcmax_instance, "V", "setUseHeuristic", useHeuristic)
    .jcall(pcmax_instance, "V", "setVerbose", verbose)

    if(!is.null(priorKnowledge)){
        .jcall(pcmax_instance, "V", "setKnowledge", priorKnowledge)
    }

    params <- c(params, continuous = continuous)
    params <- c(params, depth = as.integer(depth))
   # params <- c(params, aggressivelyPreventCycles = as.logical(aggressivelyPreventCycles))
   # params <- c(params, falseDiscoveryRate = as.logical(falseDiscoveryRate))
    params <- c(params, significance = significance)
   # params <- c(params, maxPathLength = as.integer(maxPathLength))
   # params <- c(params, useHeuristic = as.logical(useHeuristic))
    params <- c(params, verbose = as.logical(verbose))

    if(!is.null(priorKnowledge)){
        params <- c(params, prior = priorKnowledge)
    }
    pcmax$parameters <- params


    cat("Graph Parameters:\n")
    cat("continuous = ", continuous, "\n")
    cat("depth = ", as.integer(depth),"\n")
   # cat("aggressivelyPreventCycles = ", aggressivelyPreventCycles, "\n")
   # cat("falseDiscoveryRate = ", falseDiscoveryRate, "\n")
    cat("significance = ", as.integer(significance), "\n")
    cat("verbose = ", verbose, "\n")
   # cat("maxPathLength = ", as.integer(maxPathLength), "\n")
   # cat("useHeuristic = ", useHeuristic, "\n")

    # Search
    tetrad_graph <- .jcall(pcmax_instance, "Ledu/cmu/tetrad/graph/Graph;",
        "search")

    V <- extractTetradNodes(tetrad_graph)

    pcmax$nodes <- V

    # extract edges
    pcmax_edges <- extractTetradEdges(tetrad_graph)

    pcmax$edges <- pcmax_edges

    # convert output of PC into an R object (graphNEL)
    pcmax_graphNEL = tetradPattern2graphNEL(resultGraph = tetrad_graph,
        verbose = verbose)

    pcmax$graphNEL <- pcmax_graphNEL

    return(pcmax)
}



________________________________

**Comment by chirayukong** [comment]
Thanks, Scott. :-) Anyway, the pcmax was added. If you have a chance, please check the new library out.
